plugins/destination/managed_writer.go

```diff
@@ -0,0 +1,170 @@
+					}
+					resources = make([][]interface{}, 0)
+				}
+			} else {
+				if len(resources) > 0 {
+					start := time.Now()
+					if err := p.client.WriteTableBatch(ctx, table, resources); err != nil {
+						p.logger.Err(err).Str("table", table.Name).Int("len", len(resources)).Dur("duration", time.Since(start)).Msg("failed to write last batch")
+						atomic.AddUint64(&metrics.Errors, uint64(len(resources)))
+					} else {
+						p.logger.Info().Str("table", table.Name).Int("len", len(resources)).Dur("duration", time.Since(start)).Msg("last batch written successfully")
+						atomic.AddUint64(&metrics.Writes, uint64(len(resources)))
+					}
+				}
+				return
+			}
+		case <-time.After(batchTimeout):
+			if len(resources) > 0 {
+				start := time.Now()
+				if err := p.client.WriteTableBatch(ctx, table, resources); err != nil {
+					p.logger.Err(err).Str("table", table.Name).Int("len", len(resources)).Dur("time", time.Since(start)).Msg("failed to write batch on timeout")
+					// we don't return as we need to continue until channel is closed otherwise there will be a deadlock
+					atomic.AddUint64(&metrics.Errors, uint64(len(resources)))
+				} else {
+					p.logger.Info().Str("table", table.Name).Int("len", len(resources)).Dur("time", time.Since(start)).Msg("batch written successfully on timeout")
+					atomic.AddUint64(&metrics.Writes, uint64(len(resources)))
+				}
+				resources = make([][]interface{}, 0)
+			}
+		case done := <-flush:
+			if len(resources) > 0 {
+				start := time.Now()
+				if err := p.client.WriteTableBatch(ctx, table, resources); err != nil {
+					p.logger.Err(err).Str("table", table.Name).Int("len", len(resources)).Dur("time", time.Since(start)).Msg("failed to write batch on timeout")
+					// we don't return as we need to continue until channel is closed otherwise there will be a deadlock
+					atomic.AddUint64(&metrics.Errors, uint64(len(resources)))
+				} else {
+					p.logger.Info().Str("table", table.Name).Int("len", len(resources)).Dur("time", time.Since(start)).Msg("batch written successfully on timeout")
+					atomic.AddUint64(&metrics.Writes, uint64(len(resources)))
+				}
+				resources = make([][]interface{}, 0)
+				done <- true
+			}
+		}
+	}
+}
+
+func (p *Plugin) writeManagedTableBatch(ctx context.Context, tables schema.Tables, sourceName string, syncTime time.Time, res <-chan schema.DestinationResource) error {
+	syncTime = syncTime.UTC()
+	SetDestinationManagedCqColumns(tables)
+
+	workers := make(map[string]*worker, len(tables))
+	metrics := &Metrics{}
+
+	p.workersLock.Lock()
+	// we call this after workers lock as we don't want to call PreWriteTableBatch concurrently
+	if err := p.client.PreWriteTableBatch(ctx, tables); err != nil {
+		p.workersLock.Unlock()
+		return err
+	}
+	for _, table := range tables.FlattenTables() {
+		table := table
+		if p.workers[table.Name] == nil {
+			ch := make(chan schema.CQTypes)
+			flush := make(chan chan bool)
+			wg := &sync.WaitGroup{}
```

Comment:
It looks like we never wait on this waitgroup?

Reply:
Yes, good catch. fixed.

Reply:
I still don't see where we're waiting 🤔

Reply:
Nevermind, I see it's a property of the worker and getting waited on lower down